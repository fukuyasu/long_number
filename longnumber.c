#include <stdio.h>
#include <stdlib.h>
#include "longnumber.h"

static int LN_DIGITS;   /* max 9 digits with unsigned int */
static unsigned long long LN_MAXNUM = 1U;

int ln_init(int digits)
{
    if (digits < 1 || 9 < digits) {
        fprintf(stderr, "Warning: %d: invalid digit size.\n", digits);
        digits = 4;
    }
    LN_DIGITS = digits;
    if (LN_MAXNUM == 1) {
        int i;
        for (i = 0; i < LN_DIGITS; i++) {
            LN_MAXNUM *= 10U;
        }
    }
    return digits;
}

LongNumber ln_create(size_t l, unsigned int x, unsigned int y)
{
    unsigned int *ln;
    ln = (unsigned int *)calloc(sizeof(unsigned int), l+2);
    if (ln == NULL) {
        fprintf(stderr, "Error: Can't allocate memory.\n");
        exit(1);
    }
    ln[0] = x;
    if (y > 1) {
        ln_div(ln, ln, y, l+1);
    }
    return ln;
}

void ln_free(LongNumber ln)
{
    if (ln != NULL) free(ln);
}

void ln_print(LongNumber c, size_t l, char *s)
{
    int i;
    size_t sp = 1, ep;

    int cols = 80;
    char *env_columns = getenv("COLUMNS");

    if (env_columns != NULL) {
        cols = atoi(env_columns);
        if (cols < 40) cols = 40;
    }

    printf("%3s %3u.", s, c[0]);
    for (i = 1; i < l; i++) {
        printf("%0*u ", LN_DIGITS, c[i]);
        if (i % ((cols-30)/(LN_DIGITS+1)) == 0 && i != l-1) {
            ep = i * LN_DIGITS;
            printf("(%4zu-%4zu)\n", sp, ep);
            printf("        ");
            sp = ep + 1;
        }
    }
    ep = (l-1) * LN_DIGITS;
    printf("(%4zu-%4zu)\n", sp, ep);
}

void ln_add(LongNumber c, LongNumber a, LongNumber b, size_t l)
{
    int i;
    int cy = 0;
    for (i = l+1; i >= 0; i--) {
        c[i] = a[i] + b[i] + cy;  /* 0 <= c[i] <= 2*LN_MAXNUM - 1 */
        if (c[i] < LN_MAXNUM) {
            cy = 0;
        } else {
            c[i] -= LN_MAXNUM;
            cy = 1;
        }
    }
}

void ln_sub(LongNumber c, LongNumber a, LongNumber b, size_t l)
{
    int i;
    int br = 0;
    for (i = l+1; i >= 0; i--) {
        if (a[i] < b[i] + br) { /* b[i] + br <= LN_MAXNUM */
            c[i] = a[i] + (LN_MAXNUM - b[i] - br);
            br = 1;
        } else {
            c[i] = a[i] - b[i] - br;
            br = 0;
        }
    }
}

void ln_div(LongNumber c, LongNumber a, unsigned int b, size_t l)
{
    int i;
    unsigned int rem = 0;
    unsigned long long d;
    for (i = 0; i <= l+1; i++){
        d = a[i] + rem * LN_MAXNUM; /* 0 <= d <= b * LN_MAXNUM - 1 */
        c[i] = d / b;               /* 0 <= c[i] <= LN_MAXNUM - 1 */
        rem = d % b;                /* 0 <= rem <= b-1 */
    }
}

int ln_eps(LongNumber a, size_t l)
{
    int i;
    for (i = 0; i <= l+1; i++) {
        if (a[i] != 0) {
            return 0;
        }
    }
    return 1;
}
