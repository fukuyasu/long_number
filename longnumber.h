#if !defined(_LONG_NUMBER_H)
#define _LONG_NUMBER_H

typedef struct {
    unsigned int int_d;
    unsigned int *frac_d;
    /* size_t frac_len; */
    /* size_t digits; */
    /* unsigned int max_value; */
} LongNumber;

int ln_init(int digits);
LongNumber *ln_create(size_t l, unsigned int x, unsigned int y);
void ln_free(LongNumber *ln);
void ln_print(LongNumber *c, size_t l, char *s);
void ln_add(LongNumber *c, LongNumber *a, LongNumber *b, size_t l);
void ln_sub(LongNumber *c, LongNumber *a, LongNumber *b, size_t l);
void ln_div(LongNumber *c, LongNumber *a, unsigned int b, size_t l);
int ln_eps(LongNumber *a, size_t l);

#endif /* !defined(_LONG_NUMBER_H) */
