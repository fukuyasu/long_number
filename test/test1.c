#include <stdio.h>
#include "longnumber.h"

void test_ln_create01(void)
{
    LongNumber *a;
    a = ln_create(10, 0, 0);
    ln_print(a, 10+2, "c01");
    ln_free(a);
}

void test_ln_create02(void)
{
    LongNumber *a;
    a = ln_create(10, 1, 3);
    ln_print(a, 10+2, "c02");
    ln_free(a);
}

void test_ln_create03(void)
{
    LongNumber *a;
    a = ln_create(10, 2, 3);
    ln_print(a, 10+2, "c03");
    ln_free(a);
}

void test_ln_add01(void)
{
    LongNumber *a, *b;
    a = ln_create(10, 0, 0);
    b = ln_create(10, 0, 0);
    ln_add(b, b, a, 10);
    ln_print(b, 10+2, "a01");
    ln_free(a);
    ln_free(b);
}

void test_ln_add02(void)
{
    LongNumber *a, *b;
    a = ln_create(10, 1, 3);
    b = ln_create(10, 2, 3);
    ln_add(b, b, a, 10);
    ln_print(b, 10+2, "a02");
    ln_free(a);
    ln_free(b);
}

void test_ln_add03(void)
{
    LongNumber *a, *b;
    a = ln_create(10, 2, 3);
    b = ln_create(10, 2, 3);
    ln_add(b, b, a, 10);
    ln_print(b, 10+2, "a03");
    ln_free(a);
    ln_free(b);
}

int main(void)
{
    ln_init(4);
    test_ln_create01();
    test_ln_create02();
    test_ln_create03();
    test_ln_add01();
    test_ln_add02();
    test_ln_add03();

    ln_init(1);
    test_ln_create01();

    return 0;
}
