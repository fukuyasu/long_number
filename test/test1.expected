c01   0.0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 (   1-  40)
        0000 (  41-  44)
c02   0.3333 3333 3333 3333 3333 3333 3333 3333 3333 3333 (   1-  40)
        3333 (  41-  44)
c03   0.6666 6666 6666 6666 6666 6666 6666 6666 6666 6666 (   1-  40)
        6666 (  41-  44)
a01   0.0000 0000 0000 0000 0000 0000 0000 0000 0000 0000 (   1-  40)
        0000 (  41-  44)
a02   0.9999 9999 9999 9999 9999 9999 9999 9999 9999 9999 (   1-  40)
        9999 (  41-  44)
a03   1.3333 3333 3333 3333 3333 3333 3333 3333 3333 3333 (   1-  40)
        3332 (  41-  44)
c01   0.0 0 0 0 0 0 0 0 0 0 0 (   1-  11)
